### ISSUE TITLE
Consider default `as_factor` conversion for `geo_type`

### ISSUE THREAD

**Issue by dsweber2** [cmu-delphi/epiprocess#501]
Working with some [one hot encoding in recipes](https://recipes.tidymodels.org/reference/step_string2factor.html) and they were suggesting converting to factors beforehand. Given this use case, it may be worth just default converting to a factor as we create the `epi_df` or `archive`. Are there any disadvantages to having a factor instead of a character? 

And as long as we're detecting the `geo_type`, we may want to include a `levels_method` with values `"infer"` (only the ones present in this tibble), `"auto"` (based on the `geo_type`) and `<vector of levels>` (i.e. custom).

**Comment by dsweber2** [comment]
on a related note, it looks like if `geo_value` is a factor, our methods to detect the `geo_type` are wrong.

**Comment by dajmcdon** [comment]
If you actually did this as a factor, then you'd need the possible levels that it could take. By default, this will be all unique geos, but that might not be a good default in this case. What are the advantages of having it as a factor (rather than creating another column that is converted to factor).

**Comment by dsweber2** [comment]
It's required for one hot encoding at a minimum. It requires the user to know to include `df["one_hot_geo"] <- df["geo_value"]`; `df <- mutate(df, one_hot_geo, as.factor))` to be able to add the geo as a 1 hot encoding. 

It's also a more accurate reflection of the state; if we're saying `geo_type="state"`, we know that takes from a specific list of potential geo_values

**Comment by dajmcdon** [comment]
It's not necessary to convert manually like that within `{tidymodels}`. For example:

``` r
library(epipredict)
library(recipes)

jhu <- case_death_rate_subset |>
  filter(time_value > "2021-08-01")

r <- epi_recipe(jhu, death_rate ~ case_rate) |>
  add_role(geo_value, new_role = "predictor")

wf <- epi_workflow(r, linear_reg()) |> fit(jhu)
extract_fit_engine(wf)
#> 
#> Call:
#> stats::lm(formula = ..y ~ ., data = data)
#> 
#> Coefficients:
#> (Intercept)  geo_valueal  geo_valuear  geo_valueas  geo_valueaz  geo_valueca  
#>    0.251820     0.253036     0.251651    -0.252320     0.117198    -0.134668  
#> geo_valueco  geo_valuect  geo_valuedc  geo_valuede  geo_valuefl  geo_valuega  
#>   -0.030454    -0.196406    -0.330723    -0.066594     0.294937     0.198525  
#> geo_valuegu  geo_valuehi  geo_valueia  geo_valueid  geo_valueil  geo_valuein  
#>    0.098512    -0.108438    -0.062818     0.285946    -0.166597     0.053469  
#> geo_valueks  geo_valueky  geo_valuela  geo_valuema  geo_valuemd  geo_valueme  
#>   -0.024075     0.234814     0.166400    -0.198545    -0.157687    -0.097181  
#> geo_valuemi  geo_valuemn  geo_valuemo  geo_valuemp  geo_valuems  geo_valuemt  
#>    0.075619    -0.104585     0.266589    -0.267692     0.223687     0.290150  
#> geo_valuenc  geo_valuend  geo_valuene  geo_valuenh  geo_valuenj  geo_valuenm  
#>   -0.039770    -0.043730    -0.035038    -0.167120    -0.205679     0.019657  
#> geo_valuenv  geo_valueny  geo_valueoh  geo_valueok  geo_valueor  geo_valuepa  
#>    0.172622    -0.210508     0.175378     0.418519     0.061301     0.036794  
#> geo_valuepr  geo_valueri  geo_valuesc  geo_valuesd  geo_valuetn  geo_valuetx  
#>   -0.191080    -0.229462     0.174344    -0.085728     0.321066     0.126240  
#> geo_valueut  geo_valueva  geo_valuevi  geo_valuevt  geo_valuewa  geo_valuewi  
#>   -0.138359    -0.064920    -0.026580    -0.186618    -0.053544    -0.112827  
#> geo_valuewv  geo_valuewy    case_rate  
#>    0.391457     0.380041     0.003837
```

<sup>Created on 2024-08-13 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

Generally, string variables are automatically converted to factors (in fact _ALL_ strings are converted as factors, but internals of `{epipredict}` prevent this for the key-cols). Converting the factors to predictors depends on the engine requirements. Note that an unregularized linear model can't handle a one-hot encoding due to multicolinearity with the intercept. So this is coded as a "treatment contrast" where the first level is the "control" and so is dropped from the model. For other engines, different processing is desired. For example, `randomForest()` (the engine) prefers no encoding. The `grf_quantiles` implementation that I wrote uses one-hot by default, but see this nice [vignette](https://grf-labs.github.io/grf/articles/categorical_inputs.html) of theirs.

**Comment by dshemetov** [comment]
Seems like this is pretty clearly a downstream concern for the engine. Move to close?

**Comment by dsweber2** [comment]
The reason I opened it here is that `factor` more directly expresses our expectation of what is in that column. If it's a state, it should be from the list of states and error otherwise. As long as we're recording `geo_type`, this would encode that information more explicitly. I'm not saying this is particularly high priority atm.

**Comment by dajmcdon** [comment]
Agreed on the utility of explicitly encoding the possible values. But I suspect this would be very hard to maintain. States are "fine", but what about the other types?

**Comment by dshemetov** [comment]
The whole `geo_type` metadata feels like it drifts this package into geocoding concerns, which takes a lot of effort to get right. I think we should think twice about committing to that work.
